pkg/icingadb/schema.go

```diff
@@ -56,3 +58,109 @@ func CheckSchema(ctx context.Context, db *database.DB) error {
 
 	return nil
 }
+
+// ImportSchema performs an initial schema import in the db.
+//
+// This function starts by checking if at least a partial schema exists. If it is present, a nil error will be returned.
+// Otherwise, the schema file for the db's driver will be loaded from databaseSchemaDir and imported into db.
+//
+// The function returns a bool - true when an import was performed - and a potential error. If the error is not nil,
+// the bool should also be false. When (false, nil) is returned, a schema is already imported.
+func ImportSchema(
+	ctx context.Context,
+	db *database.DB,
+	databaseName string,
+	databaseSchemaDir string,
+) (bool, error) {
+	var tableSchema, schemaFileDirPart string
+	switch db.DriverName() {
+	case database.MySQL:
+		tableSchema = databaseName
+		schemaFileDirPart = "mysql"
+
+	case database.PostgreSQL:
+		tableSchema = "public"
+		schemaFileDirPart = "pgsql"
+
+	default:
+		return false, fmt.Errorf("unsupported database driver %q", db.DriverName())
+	}
+
+	tx, err := db.BeginTxx(ctx, nil)
+	if err != nil {
+		return false, errors.Wrap(err, "can't start database transaction")
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	var schemaTablesCount int
+	err = retry.WithBackoff(
+		ctx,
+		func(ctx context.Context) (err error) {
+			query := tx.Rebind("SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA=? AND TABLE_NAME='icingadb_schema'")
+			err = tx.QueryRowxContext(ctx, query, tableSchema).Scan(&schemaTablesCount)
+			if err != nil {
+				err = database.CantPerformQuery(err, query)
+			}
+			return
+		},
+		retry.Retryable,
+		backoff.NewExponentialWithJitter(128*time.Millisecond, 1*time.Minute),
+		db.GetDefaultRetrySettings())
+	if err != nil {
+		return false, errors.Wrap(err, "can't check if database has the schema applied")
+	}
+	if schemaTablesCount > 0 {
+		// icingadb_schema table exists; schema is at least partially applied
+		return false, nil
+	}
+
+	schemaFile := path.Join(databaseSchemaDir, schemaFileDirPart, "schema.sql")
+	schema, err := os.ReadFile(schemaFile)
+	if err != nil {
+		return false, errors.Wrapf(err, "can't open schema file %q", schema)
+	}
+
+	switch db.DriverName() {
+	case database.MySQL:
+		for _, query := range database.MysqlSplitStatements(string(schema)) {
+			err := retry.WithBackoff(
+				ctx,
+				func(ctx context.Context) (err error) {
+					_, err = tx.ExecContext(ctx, query)
+					if err != nil {
+						err = database.CantPerformQuery(err, query)
+					}
+					return
+				},
```

Comment:
I would always prefer the `if err := ...; err != nil { return err } ... return nil` pattern:

```suggestion
				func(ctx context.Context) error {
					if _, err := tx.ExecContext(ctx, query); err != nil {
						return database.CantPerformQuery(err, query) 
					}
					return nil
				},
```

Reply:
I have tried to keep the new function consistent with the other `CheckSchema` function. So I have changed it there as well :)